File: ports/go/core/timers.go
package core

import (
	"encoding/json"
	"sort"
	"sync"
)

// Timer is a scheduled event: when logical/processing time reaches FireAt, Payload is
// fired back into the runtime as a turn. ID is unique — re-scheduling the same id
// replaces the pending timer.
type Timer struct {
	ID      string
	FireAt  int64
	Payload Event
}

// TimerService is portable timers — the counterpart of Flink's TimerService, a Pekko
// scheduler, or a Temporal timer. Time is logical: callers advance the clock with
// AdvanceTo(now) and get back the due timers (so tests are deterministic); a real-time
// driver simply calls AdvanceTo(now) on a tick. Powers SLAs, escalate-after-N, retries,
// scheduled follow-ups, and CEP "within" expiry. Engine adapters may replace this with a
// native timer service.
type TimerService interface {
	// Schedule (or replace, by id) a timer to fire payload at fireAt. A replaced timer
	// takes the new schedule order.
	Schedule(id string, fireAt int64, payload Event)
	// Cancel a pending timer; returns true if one was removed.
	Cancel(id string) bool
	// AdvanceTo removes and returns all timers due at now (FireAt <= now), ascending by
	// FireAt with schedule order as the stable tie-break.
	AdvanceTo(now int64) []Timer
	// NextDeadline is the earliest pending FireAt, or ok=false if none are pending.
	NextDeadline() (int64, bool)
}

// scheduled pairs a timer with its monotonic insertion sequence so equal deadlines keep
// schedule order as a stable tie-break.
type scheduled struct {
	timer Timer
	seq   int64
}

// InMemoryTimerService is process-local timers backed by a map keyed by id plus a
// monotonic sequence; due timers come out ascending by FireAt, with schedule order as the
// stable tie-break. Goroutine-safe via a mutex.
type InMemoryTimerService struct {
	mu     sync.Mutex
	timers map[string]scheduled
	seq    int64
}

// NewInMemoryTimerService builds an empty in-memory timer service.
func NewInMemoryTimerService() *InMemoryTimerService {
	return &InMemoryTimerService{timers: map[string]scheduled{}}
}

// Schedule inserts or replaces by id; a replaced timer takes a fresh sequence so it sorts
// after timers scheduled before this call (new schedule order).
func (s *InMemoryTimerService) Schedule(id string, fireAt int64, payload Event) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.timers[id] = scheduled{timer: Timer{ID: id, FireAt: fireAt, Payload: payload}, seq: s.seq}
	s.seq++
}

// Cancel removes a pending timer, returning true if one was present.
func (s *InMemoryTimerService) Cancel(id string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	if _, ok := s.timers[id]; ok {
		delete(s.timers, id)
		return true
	}
	return false
}

// AdvanceTo removes and returns timers with FireAt <= now, ascending by FireAt then
// schedule order.
func (s *InMemoryTimerService) AdvanceTo(now int64) []Timer {
	s.mu.Lock()
	defer s.mu.Unlock()
	var due []scheduled
	for _, sc := range s.timers {
		if sc.timer.FireAt <= now {
			due = append(due, sc)
		}
	}
	sortScheduled(due)
	out := make([]Timer, 0, len(due))
	for _, sc := range due {
		out = append(out, sc.timer)
		delete(s.timers, sc.timer.ID)
	}
	return out
}

// NextDeadline returns the earliest pending FireAt, or ok=false when none are pending.
func (s *InMemoryTimerService) NextDeadline() (int64, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if len(s.timers) == 0 {
		return 0, false
	}
	var min int64
	first := true
	for _, sc := range s.timers {
		if first || sc.timer.FireAt < min {
			min = sc.timer.FireAt
			first = false
		}
	}
	return min, true
}

// pending returns a snapshot of pending timers in schedule order (for durable persistence).
func (s *InMemoryTimerService) pending() []Timer {
	s.mu.Lock()
	defer s.mu.Unlock()
	all := make([]scheduled, 0, len(s.timers))
	for _, sc := range s.timers {
		all = append(all, sc)
	}
	sort.SliceStable(all, func(i, j int) bool { return all[i].seq < all[j].seq })
	out := make([]Timer, 0, len(all))
	for _, sc := range all {
		out = append(out, sc.timer)
	}
	return out
}

// restoreAll re-inserts timers, preserving the given schedule order via fresh sequences.
func (s *InMemoryTimerService) restoreAll(restored []Timer) {
	s.mu.Lock()
	defer s.mu.Unlock()
	for _, t := range restored {
		s.timers[t.ID] = scheduled{timer: t, seq: s.seq}
		s.seq++
	}
}

// sortScheduled orders by FireAt ascending, then by insertion sequence (schedule order).
func sortScheduled(xs []scheduled) {
	sort.SliceStable(xs, func(i, j int) bool {
		if xs[i].timer.FireAt != xs[j].timer.FireAt {
			return xs[i].timer.FireAt < xs[j].timer.FireAt
		}
		return xs[i].seq < xs[j].seq
	})
}

// DurableTimerService persists its pending set through a KeyedStateStore, so timers
// survive a restart (with a durable store backing). The pending set is written to one
// scalar slot as JSON; Restore reloads it. Schedule/Cancel/AdvanceTo keep the slot
// current.
//
// The serialization is per-process self-consistent (never read across languages), so each
// core port may encode it idiomatically; behaviour — "a scheduled timer survives restore
// and fires" — is what stays at parity.
type DurableTimerService struct {
	mu       sync.Mutex
	delegate *InMemoryTimerService
	store    KeyedStateStore
	slotKey  string
	slotName string
}

// NewDurableTimerService wraps an in-memory service, persisting to the default slot
// (key "__timers__", name "pending").
func NewDurableTimerService(store KeyedStateStore) *DurableTimerService {
	return NewDurableTimerServiceWithSlot(store, "__timers__", "pending")
}

// NewDurableTimerServiceWithSlot lets callers pick the persistence slot key/name.
func NewDurableTimerServiceWithSlot(store KeyedStateStore, slotKey, slotName string) *DurableTimerService {
	return &DurableTimerService{
		delegate: NewInMemoryTimerService(),
		store:    store,
		slotKey:  slotKey,
		slotName: slotName,
	}
}

// Schedule schedules (or replaces) a timer and persists the pending set.
func (d *DurableTimerService) Schedule(id string, fireAt int64, payload Event) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.delegate.Schedule(id, fireAt, payload)
	d.persist()
}

// Cancel removes a pending timer, persisting if one was removed.
func (d *DurableTimerService) Cancel(id string) bool {
	d.mu.Lock()
	defer d.mu.Unlock()
	removed := d.delegate.Cancel(id)
	if removed {
		d.persist()
	}
	return removed
}

// AdvanceTo returns due timers, persisting the pending set if any fired.
func (d *DurableTimerService) AdvanceTo(now int64) []Timer {
	d.mu.Lock()
	defer d.mu.Unlock()
	due := d.delegate.AdvanceTo(now)
	if len(due) > 0 {
		d.persist()
	}
	return due
}

// NextDeadline delegates to the in-memory service.
func (d *DurableTimerService) NextDeadline() (int64, bool) {
	d.mu.Lock()
	defer d.mu.Unlock()
	return d.delegate.NextDeadline()
}

// Restore reloads pending timers from the store into this service (call after a restart).
func (d *DurableTimerService) Restore() {
	d.mu.Lock()
	defer d.mu.Unlock()
	raw, ok := d.store.Get(d.slotKey, d.slotName)
	if !ok || raw == nil {
		return
	}
	s, ok := raw.(string)
	if !ok {
		return
	}
	d.delegate.restoreAll(decodeTimers(s))
}

// persist writes the current pending set to the store slot. Caller holds d.mu.
func (d *DurableTimerService) persist() {
	d.store.Put(d.slotKey, d.slotName, encodeTimers(d.delegate.pending()))
}

// persistedTimer is the JSON wire shape for one timer (per-process; never cross-language).
type persistedTimer struct {
	ID             string `json:"id"`
	FireAt         int64  `json:"fireAt"`
	ConversationID string `json:"conversationID"`
	UserID         string `json:"userID"`
	Text           string `json:"text"`
}

// encodeTimers serializes pending timers to a JSON array string.
func encodeTimers(timers []Timer) string {
	out := make([]persistedTimer, 0, len(timers))
	for _, t := range timers {
		out = append(out, persistedTimer{
			ID:             t.ID,
			FireAt:         t.FireAt,
			ConversationID: t.Payload.ConversationID,
			UserID:         t.Payload.UserID,
			Text:           t.Payload.Text,
		})
	}
	b, err := json.Marshal(out)
	if err != nil {
		return "[]"
	}
	return string(b)
}

// decodeTimers parses the JSON array string back into Timers (metadata is not persisted).
func decodeTimers(s string) []Timer {
	var in []persistedTimer
	if err := json.Unmarshal([]byte(s), &in); err != nil {
		return nil
	}
	out := make([]Timer, 0, len(in))
	for _, p := range in {
		out = append(out, Timer{
			ID:     p.ID,
			FireAt: p.FireAt,
			Payload: Event{
				ConversationID: p.ConversationID,
				UserID:         p.UserID,
				Text:           p.Text,
			},
		})
	}
	return out
}

File: ports/pyagentic/pyagentic/timers.py
"""Portable timers — the Python counterpart of the Java ``org.jagentic.core.timers``.

The portable analogue of Flink's ``TimerService`` / a Pekko scheduler / a Temporal
timer. Time is **logical**: callers advance the clock with :meth:`TimerService.advance_to`
and get back the due timers (so tests are deterministic); a real-time driver simply calls
``advance_to(now)`` on a tick. Powers SLAs, escalate-after-N, retries, scheduled
follow-ups, and CEP ``within`` expiry. Engine adapters may replace this entirely with a
native timer service.

``DurableTimerService`` persists its pending set through a :class:`KeyedStateStore`, so
timers survive a restart. The serialization is per-process self-consistent (it is never
read across languages), so each core port may encode it idiomatically — here JSON; the
behaviour ("a scheduled timer survives restore and fires") is what stays at parity.
"""

from __future__ import annotations

import json
from collections import OrderedDict
from dataclasses import dataclass
from threading import RLock
from typing import Dict, List, Optional, Protocol, runtime_checkable

from .core import Event
from .memory import KeyedStateStore


@dataclass(frozen=True)
class Timer:
    """A scheduled event: when logical/processing time reaches ``fire_at``, ``payload``
    is fired back into the runtime as a turn. ``id`` is unique (re-scheduling the same
    id replaces)."""

    id: str
    fire_at: int
    payload: Event


@runtime_checkable
class TimerService(Protocol):
    """Portable timers. Mirrors ``org.jagentic.core.timers.TimerService``."""

    def schedule(self, id: str, fire_at: int, payload: Event) -> None:
        """Schedule (or replace, by id) a timer to fire ``payload`` at ``fire_at``."""
        ...

    def cancel(self, id: str) -> bool:
        """Cancel a pending timer; returns True if one was removed."""
        ...

    def advance_to(self, now: int) -> List[Timer]:
        """Remove and return all timers due at ``now`` (fire_at <= now), ascending by
        fire_at then schedule order."""
        ...

    def next_deadline(self) -> Optional[int]:
        """The earliest pending deadline, or None if no timers are pending."""
        ...


class InMemoryTimerService:
    """Process-local timers backed by an insertion-ordered dict; due timers come out
    ascending by fire_at, with schedule order as the stable tie-break."""

    def __init__(self) -> None:
        # Python dicts preserve insertion order; OrderedDict makes the intent explicit
        # and gives move_to_end semantics for the replace-keeps-new-order rule.
        self._timers: "OrderedDict[str, Timer]" = OrderedDict()
        self._lock = RLock()

    def schedule(self, id: str, fire_at: int, payload: Event) -> None:
        with self._lock:
            # Re-insert so a replaced timer takes the new schedule order.
            self._timers.pop(id, None)
            self._timers[id] = Timer(id, fire_at, payload)

    def cancel(self, id: str) -> bool:
        with self._lock:
            return self._timers.pop(id, None) is not None

    def advance_to(self, now: int) -> List[Timer]:
        with self._lock:
            due = [t for t in self._timers.values() if t.fire_at <= now]
            # Stable sort -> equal fire_at keeps schedule order (the dict iteration order).
            due.sort(key=lambda t: t.fire_at)
            for t in due:
                self._timers.pop(t.id, None)
            return due

    def next_deadline(self) -> Optional[int]:
        with self._lock:
            if not self._timers:
                return None
            return min(t.fire_at for t in self._timers.values())

    def pending(self) -> List[Timer]:
        """Snapshot of pending timers in schedule order (for durable persistence)."""
        with self._lock:
            return list(self._timers.values())

    def restore_all(self, restored: List[Timer]) -> None:
        with self._lock:
            for t in restored:
                self._timers[t.id] = t


class DurableTimerService:
    """A :class:`TimerService` that persists its pending set through a
    :class:`KeyedStateStore`, so timers survive a restart (with a durable store
    backing). Pending timers are written to one scalar slot; :meth:`restore` reloads
    them. Schedule/cancel/advance keep the slot current."""

    def __init__(
        self,
        store: KeyedStateStore,
        slot_key: str = "__timers__",
        slot_name: str = "pending",
    ) -> None:
        self._delegate = InMemoryTimerService()
        self._store = store
        self._slot_key = slot_key
        self._slot_name = slot_name
        self._lock = RLock()

    def schedule(self, id: str, fire_at: int, payload: Event) -> None:
        with self._lock:
            self._delegate.schedule(id, fire_at, payload)
            self._persist()

    def cancel(self, id: str) -> bool:
        with self._lock:
            removed = self._delegate.cancel(id)
            if removed:
                self._persist()
            return removed

    def advance_to(self, now: int) -> List[Timer]:
        with self._lock:
            due = self._delegate.advance_to(now)
            if due:
                self._persist()
            return due

    def next_deadline(self) -> Optional[int]:
        with self._lock:
            return self._delegate.next_deadline()

    def restore(self) -> None:
        """Reload pending timers from the store into this service (call after a restart)."""
        with self._lock:
            raw = self._store.get(self._slot_key, self._slot_name)
            if raw is None:
                return
            self._delegate.restore_all(_decode(str(raw)))

    def pending(self) -> List[Timer]:
        with self._lock:
            return self._delegate.pending()

    def _persist(self) -> None:
        self._store.put(self._slot_key, self._slot_name, _encode(self._delegate.pending()))


# ---- serialization: JSON list of {id, fire_at, conversation_id, user_id, text} ----


def _encode(timers: List[Timer]) -> str:
    return json.dumps(
        [
            {
                "id": t.id,
                "fire_at": t.fire_at,
                "conversation_id": t.payload.conversation_id,
                "user_id": t.payload.user_id,
                "text": t.payload.text,
            }
            for t in timers
        ]
    )


def _decode(s: str) -> List[Timer]:
    out: List[Timer] = []
    for rec in json.loads(s):
        event = Event(
            conversation_id=rec["conversation_id"],
            text=rec["text"],
            user_id=rec.get("user_id", "anonymous"),
        )
        out.append(Timer(rec["id"], int(rec["fire_at"]), event))
    return out
